### DISCUSSION TITLE
Modifying multiple commits at once, e.g. deleting a file from each

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#4584] [Q&A]
Is there a way to apply the same change to multiple commits at once, e.g. to delete a file from each commit in some revset?

**Comment by martinvonz** [comment]
No. I've considered commands for some of that. Deleting paths is easy. Adding a file with constant content across a bunch of commits is also easy. Other cases seem harder. It's the UX that's hard, not the implementation. We should probably collect some use cases first.

**Reply by arxanas** [reply]
It's a use-case for `jj run`/`jj fix`/whatever. I've used git-branchless's fix command in this specific case to avoid (effectively resolve) conflicts when trying to remove a file from many commits. I think the same workflow applies to jj.

**Reply by PhilipMetzger** [reply]
Agreed, it basically should be `jj run 'rm <file>' -r <revs>`.

**Reply by BatmanAoD** [reply]
In the particular instance I ran into today, I found that I had accidentally committed new files to my branch quite a few commits ago. In a similar instance that I run into frequently, my [unit tests modify tracked files](https://github.com/martinvonz/jj/discussions/4587) (...which, yes, I would like to fix). In both of these cases, I essentially just want to restore the state that exists in some other commit, so if `restore` supported multiple `--to` targets, that would be mostly sufficient.

There's one other wrinkle, though. As far as I can tell,  `restore --from <foo> --to <bar>` will delete added files, but `restore --from <foo> --to <bar> <PATH>` will not; so even with multiple `--to` targets, I wouldn't have been able to delete the specific files that way.

**Reply by ilyagr** [reply]
> There's one other wrinkle, though. As far as I can tell, restore --from <foo> --to <bar> will delete added files, but restore --from <foo> --to <bar> <PATH> will not

If you can reproduce this, it sounds like a bug or at the very least a missing feature to me: either the default behavior should allow deleting files, or there should be an option to allow it. In your example, is PATH a directory with multiple files or just one file?

**Reply by BatmanAoD** [reply]
It appears I was mistaken, sorry; I had dome some quick testing of this but I guess I got confused somewhere. Looks like `restore --from <foo> --to <bar> <PATH>` _does_ delete `<PATH>` if it's missing from `<foo>`.

**Reply by ilyagr** [reply]
Good to know, thanks for checking! And if you can find a case where it doesn't work, let us know.

**Comment by ilyagr** [comment]
I will close this discussion in favor of https://github.com/martinvonz/jj/discussions/4757 (choosing more or less randomly)
